lib/spack/spack/cmd/install.py

```diff
@@ -385,13 +397,17 @@ def install(parser, args, **kwargs):
                 if not answer:
                     tty.die('Reinstallation aborted.')
 
-            for abstract, concrete in zip(abstract_specs, specs):
-                if concrete in installed:
-                    with fs.replace_directory_transaction(concrete.prefix):
-                        install_spec(args, kwargs, abstract, concrete)
-                else:
-                    install_spec(args, kwargs, abstract, concrete)
-
+            if any([concrete in installed for concrete in specs]):
+                # Preserve ordering if force replacement of any installed specs
+                tty.debug('Performing serial installs due to forced '
+                          'replacement of installed concrete specs.')
+                for abstract, concrete in zip(abstract_specs, specs):
+                    if concrete in installed:
+                        with fs.replace_directory_transaction(concrete.prefix):
+                            install_specs(args, kwargs, [(abstract, concrete)])
+                    else:
+                        install_specs(args, kwargs, [(abstract, concrete)])
```

Comment:
I think this can all be replaced by the logic in #15256 . That will avoid having any serial installs

Reply:
Thanks for the PR number.  I held off changing this section after we talked about this before so I can integrate that code.

Reply:
Merged with develop to pick this up.

Reply:
> > I too noticed issues with a multi-process build of the simple, two-compiler installation environment as noted in an update I made to this PRs description. I will look into this more closely.
> > It is important to keep in mind that parallelism is effectively limited by the maximum number of packages that can be built at the same time, which itself is limited by the maximum number of packages with **no** uninstalled dependencies.
> > IIRC the two versions of `gcc` being installed shared common dependencies in their DAGS on the same versions of at least a couple of packages. You can see the overlap if you compare the output of `spack spec gcc@9.3.0` to that of `spack spec gcc@10.1.0`.
> 
> I'm aware that there's a lot of overlap, but I somehow could not get more than one process building any package at all. I just tested it by simply adding the 6-8 most recent versions of CMake to an environment (once dependencies are installed, these should all be orthogonal), and using just
> 
> ```
> spack -e debug_environment install -j 6
> ```
> 
> in 4 terminals. This has one process installing one CMake after another, while my other 3 keep saying that it is `in progress by another process`. _Maybe I'm using the command wrong?_
> 
> I added a breakpoint after re-queuing tasks if a lock cannot be obtained (around line 1540 in `installer.py`), and it seems to me that the same task is popped from the heap, which looks like:
> 
> ```
> > /gpfs/bbp.cscs.ch/home/matwolf/work/spack-origin/lib/spack/spack/installer.py(1542)install()
> -> pdb.set_trace()
> (Pdb) pprint.pprint([(t[0], t[1].pkg_id) for t in self.build_pq if t[1].status != STATUS_REMOVED])
> [((0, 13), 'cmake-3.16.5-y2q7fnh5piydazslo44gwze526kmavka'),
>  ((0, 14), 'cmake-3.16.2-lcrrvvf2ghljvwu3flrssjzdzsvvaxn3'),
>  ((0, 15), 'cmake-3.16.1-ccldrcnz7mjguukpejljyq2lpq6vvxsi')]
> ```
> 
> and where `cmake@3.16.5` is re-queued after the lock fails, with the same priority and build sequence counter. Shouldn't the key of the heap account for lock fails to bump the priority (or other key component) up a bit and allow to rotate through all tasks with the same priority? It may well be that I'm just a little too dense for this, though.

Thanks for posting the issue.  I'll look at this more once I merge to develop.